[feature] 카카오 로그인을 통한 AccessToken 발급

sanhee opened the pull request:
https://github.com/Pin-Tree/backend/issues/3 

해당 기능은 이미 백엔드 플레이그라운드 레포지토리에서 연습하면서 PR 리뷰를 거쳤기 때문에,

상세한 PR 내용은 생략하겠습니다.

시간내어 리뷰해주셔서 감사합니다!

pbg0205 committed:
feat : LoginController 추가

pbg0205 committed:
feat : LoginService 생성

pbg0205 committed:
feat : OauthServiceFactory 생성

pbg0205 committed:
feat : OauthService 생성

pbg0205 committed:
feat : GoogleOauthService 생성

pbg0205 committed:
feat : KakaoOauthService 생성

pbg0205 committed:
feat : NaverOauthService 생성

pbg0205 committed:
feat : MemberRepsitory 생성

pbg0205 committed:
feat : MemoryMemberRepository 생성

pbg0205 committed:
feat : MemberRepository save() 로직 추가

pbg0205 committed:
feat : AccessTokenResponse interface 생성

pbg0205 committed:
feat : OauthServiceFactory 생성

pbg0205 committed:
feat : OauthService interface 생성

pbg0205 committed:
feat : GoogleOauthService 생성

pbg0205 committed:
feat : KakaoOauthService 생성

pbg0205 committed:
feat : NaverOauthService 생성

pbg0205 committed:
feat : Login 관련 Component 의존성 추가

sanhee committed:
feat : WebClient Config 클래스 정의

sanhee committed:
feat : DTO 정의 (AuthRequest, AuthResponse, KaKaoAccessTokenResponse)

sanhee committed:
feat : KaKao 엑세스 토큰 발급 로직 구현

sanhee committed:
feat : KaKao 엑세스 토큰 발급 컨트롤러 구현

sanhee committed:
feat :  webflux 의존성 추가 및 kakao oauth용 properties 추가

sanhee committed:
feat :  KaKao 인증 리다이렉트 컨트롤러 정의

pbg0205 commented on pintree/src/main/java/com/trio/pintree/WebConfig.java:
```diff
@@ -0,0 +1,38 @@
+@Configuration
+public class WebConfig implements WebMvcConfigurer {
+
+    @Bean
+    public HttpClient httpClient() {
+        return HttpClient.create()
+                         .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 5000)
+                         .responseTimeout(Duration.ofMillis(5000))
+                         .doOnConnected(
+                                 conn -> conn.addHandlerLast(new ReadTimeoutHandler(5000, TimeUnit.MILLISECONDS))
+                                             .addHandlerLast(new WriteTimeoutHandler(5000, TimeUnit.MILLISECONDS))
+                         );
+    }
```
이 내용은 요청 시간 설정하는 부분이라 제거하고 나중에 지정해도 좋을 것 같아요!

sanhee replied:
https://github.com/Pin-Tree/backend/pull/6/commits/d86b1b4b9559e268fba16fb2b783fa15bfc25c4f 에서 수정했습니다.

pbg0205 commented on pintree/src/main/java/com/trio/pintree/WebConfig.java:
```diff
@@ -0,0 +1,38 @@
+                         );
+    }
+
+    @Bean
+    public WebClient webClient() {
+        return WebClient.builder()
+                        .clientConnector(new ReactorClientHttpConnector(httpClient()))
+                        .build();
+    }
```
윗 설정을 제거하면 clientConnector는 설정 안해주셔도 됩니다!
```java
    @Bean
    public WebClient webClient() {
        return WebClient.builder()
                        .build();
    }
```

sanhee replied:
https://github.com/Pin-Tree/backend/pull/6/commits/d86b1b4b9559e268fba16fb2b783fa15bfc25c4f
위에서 답변 드렸습니다!

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/component/OauthServiceFactory.java:
```diff
@@ -1,13 +1,25 @@
+    private final static String KAKAO_OAUTH_SERVICE_NAME = "kakaoOauthService";
     private final Map<String, OauthService> oauthServiceMap;
+
+    public OauthServiceFactory(Map<String, OauthService> oauthServiceMap) {
+        this.oauthServiceMap = oauthServiceMap;
+        log.debug("oauthServiceMap : {}", oauthServiceMap);
+    }
```
`@RequiredArgsConstructor`로 변경해주셔도 좋을 것 같아요!

sanhee replied:
log를 남기려고 생성자를 사용하긴 했었는데, 불필요한 것 같네요 ㅎㅎ
반영했습니다.

sanhee replied:
https://github.com/Pin-Tree/backend/pull/6/commits/72613f917ee6d776e74b6446e9b598ea8aebfbdf

에서 수정했습니다!

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/service/KakaoOauthService.java:
```diff
@@ -2,19 +2,104 @@
+
+    public KakaoOauthService(MemberRepository memberRepository,
+                             WebClient webClient,
+                             @Value("${oauth.kakao.access.token.uri}") String accessTokenUri,
+                             @Value("${oauth.kakao.get.info.uri}") String getInfoUrl,
+                             @Value("${oauth.kakao.grant.type}") String grantType,
+                             @Value("${oauth.kakao.redirect.uri}") String redirectUrl,
+                             @Value("${oauth.kakao.client.id}") String clientId,
+                             @Value("${oauth.kakao.secret.key}") String secretKey) {
```
`@ConfigurationProperties`으로 KakaoOauthProperties 클래스를 생성하면 해당 부분이 깔끔해집니다!
아래 사용법에 대해 첨부해놓을게요!
https://dune-store-e37.notion.site/ConfigurationProperties-1777ceef23034070828a461f2ff82406

sanhee replied:
https://github.com/Pin-Tree/backend/pull/6/commits/03be23218bd387693373046df424101ce0afc5d0

서비스에서 사용하던 뚱뚱한 필드들이 다이어트 됐네요!

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/controller/LoginController.java:
```diff
@@ -1,11 +1,42 @@
     private final LoginService loginService;
+
+    @GetMapping("/kakao")
+    public ResponseEntity<AuthResponse> auth(@RequestBody AuthRequest authRequest) {
```
요청이 안됐던 부분 확인해보니 code를 못받아와서 발생하는 오류였어요!
`String code`로 변경해주시면 정상 작동합니다 👍 👍

sanhee replied:
제가 GET 메서드인데, BODY에 code를 심어서 보내다보니 이런 불상사가 생긴 것 같습니다.
모두가 오해없이 사용할 수 있도록, GET의 관례적인 규칙인 Body를 사용하지 않고 설계를 해야겠네요!
꼼꼼히 리뷰해주셔서 감사합니다~!
https://github.com/Pin-Tree/backend/pull/6/commits/d393595087008ea78ef4edcaf7ddfe80ad6d73bc

pbg0205 commented on pintree/src/main/java/com/trio/pintree/login/controller/LoginController.java:
```diff
@@ -1,11 +1,42 @@
+    public ResponseEntity<AuthResponse> auth(@RequestBody AuthRequest authRequest) {
+        String code = authRequest.getCode();
+
+        OauthServiceFactory oauthServiceFactory = loginService.getOauthServiceFactory();
+        KakaoOauthService kakaoOauthService = oauthServiceFactory.getKaKaoOauthService();
+
+        KaKaoAccessTokenResponse accessTokenResponse = kakaoOauthService.issueAccessToken(code);
```
이 부분은 한줄로 줄일 수 있을거 같아요!
```java

KaKaoAccessTokenResponse accessTokenResponse = oauthServiceFactory.issueKakaoAccessToken(code);

```

sanhee replied:
https://github.com/Pin-Tree/backend/pull/6/commits/e8b28968cc445d640e774e6e83be9786d953be76
와 같이 수정완료했습니다!

pbg0205 reviewed (approved):
하루만에 WebClient 사용하시다니 대단해요... 👍
전반적으로 잘 작성해주셨고 코멘트 첨부한 부분만 참고 부탁드릴게요~

sanhee committed:
fix: 아직까지 불필요한 HttpClient 설정 제거

sanhee committed:
refactor: 생성자 제거하고 롬복 생성자 애노테이션 사용

sanhee committed:
refactor : OauthService에서 Property 필드를 줄이기 위해, @ConfigurationProperties 사용

sanhee committed:
fix : LoginController에서 code를 body가 아닌 쿼리스트링으로 받도록 수정

GET 메서드는 body를 포함해서 전송을 할 수는 있지만, 관례상 body를 사용하지 않는 경우가 많으므로 이를 지키기 위해서 쿼리스트링 형식으로 변경했습니다.

sanhee committed:
refactor : 엑세스 토큰 발급 컨트롤러 디미터 법칙을 고려하여 개선 (커밋 상세정보 확인)

+ 메서드명 auth -> issueAccessToken 으로 변경
+ OauthServiceFactory에서 accessToken을 가져올 수 있는 메서드 정의
+ AccessTokenResponse 인터페이스에 getAccessToken() 추가

sanhee commented:
@pbg0205 쿠퍼!! 리뷰 내용 반영 완료하여, `re-request` 드립니다!

큰 변경사항은 없고, AccessTokenResponse 인터페이스에서 엑세스 토큰을 가져올 수 있도록 추상 메서드 한 개 추가 했습니다~!!
https://github.com/Pin-Tree/backend/pull/6/commits/e8b28968cc445d640e774e6e83be9786d953be76

```java
public interface AccessTokenResponse {
    String getAccessToken();
}
```

pbg0205 reviewed (approved):
수고 많았어요. 노을!!! 👍 👍 👍
해당 내용 반영해서 머지하도록 할게요 👏👏👏
